Packages/com.unity.inputsystem/InputSystem/Plugins/WebGL/WebGLSupport.cs

```diff
@@ -10,22 +14,126 @@ namespace UnityEngine.InputSystem.WebGL
 #endif
     static class WebGLSupport
     {
+        private const string InterfaceName = "WebGL";
         public static void Initialize()
         {
             // We only turn gamepads with the "standard" mapping into actual Gamepads.
             InputSystem.RegisterLayout<WebGLGamepad>(
                 matches: new InputDeviceMatcher()
-                    .WithInterface("WebGL")
+                    .WithInterface(InterfaceName)
                     .WithDeviceClass("Gamepad")
                     .WithCapability("mapping", "standard"));
 
-            ////TODO: add a layout builder for this that dynamically looks at the axis and button count
-            // For all other WebGL "gamepads", we don't know the button and axis mappings so we turn
-            // them into joysticks.
-            InputSystem.RegisterLayout<WebGLJoystick>(
-                matches: new InputDeviceMatcher()
-                    .WithInterface("WebGL")
-                    .WithDeviceClass("Gamepad"));
+            InputSystem.onFindLayoutForDevice += OnFindLayoutForDevice;
+        }
+
+        internal static string OnFindLayoutForDevice(int deviceId, ref InputDeviceDescription description,
+            string matchedLayout, IInputRuntime runtime)
+        {
+            // If the device isn't a WebGL device, we're not interested.
+            if (string.Compare(description.interfaceName, InterfaceName, StringComparison.InvariantCultureIgnoreCase) != 0)
+                return null;
+
+            var deviceMatcher = InputDeviceMatcher.FromDeviceDescription(description);
+
+            var layout = new WebGLLayoutBuilder {capabilities = WebGLDeviceCapabilities.FromJson(description.capabilities)};
+            InputSystem.RegisterLayoutBuilder(() => layout.Build(),
+                description.product, null, deviceMatcher);
+
+            return description.product;
+        }
```

Comment:
Wondering, how does the fallback logic work now? Previously, the matcher for gamepads with standard mappings was "more specific" (i.e. matched on more fields) than the one for joystick, so it took precedence. Now, the fallback is hooked into onFindLayoutForDevice and it seems to kick in for every WebGL device so it seems like it is overriding the WebGLGamepad case. But I assume the tests are passing so there must be something I'm not seeing.

Reply:
The answer is quite simple: It didn't (and the test was failing). Fixed.